File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/sieve/SieveDisplay.java
package wraith.fabricaeexnihilo.compatibility.rei.sieve;

import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import net.minecraft.fluid.Fluids;
import wraith.fabricaeexnihilo.compatibility.recipeviewer.SieveRecipeKey;
import wraith.fabricaeexnihilo.compatibility.recipeviewer.SieveRecipeOutputs;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SieveDisplay implements Display {
    public final boolean waterlogged;
    public final EntryIngredient input;
    public final EntryIngredient mesh;
    public final Map<EntryIngredient, List<Double>> outputs;

    public SieveDisplay(SieveRecipeKey key, SieveRecipeOutputs outputs) {
        this.waterlogged = key.waterlogged();
        this.input = EntryIngredients.ofIngredient(key.input());
        this.mesh = EntryIngredients.of(key.mesh());
        this.outputs = outputs.outputs()
                .entries()
                .stream()
                .map(entry -> Map.entry(EntryIngredients.of(entry.getKey()), entry.getValue()))
                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    @Override
    public CategoryIdentifier<?> getCategoryIdentifier() {
        return PluginEntry.SIFTING;
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return waterlogged ? List.of(input, mesh, EntryIngredients.of(Fluids.WATER)) : List.of(input, mesh);
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.copyOf(outputs.keySet());
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/sieve/SieveDisplay.kt
package exnihilofabrico.compatibility.rei.sieve

import exnihilofabrico.api.recipes.SieveRecipe
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.modules.ModBlocks
import exnihilofabrico.util.asREIEntry
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeDisplay

class SieveDisplay(val recipe: SieveRecipe): RecipeDisplay {
    override fun getRecipeCategory() = PluginEntry.SIEVE

    override fun getOutputEntries() = recipe.loot.map { it.stack.asREIEntry() }.toMutableList()
    override fun getInputEntries(): List<List<EntryStack>> {
        val sievable = recipe.sievable.asREIEntries()
        val mesh = recipe.mesh.asREIEntries()
        val buckets = recipe.fluid.asREIEntries()
        val sieves = ModBlocks.SIEVES.values.map { it.asREIEntry() }.toMutableList()

        return mutableListOf(sievable, mesh, buckets, sieves)
    }

}